Project 1 Report: A backend translates Bril into Java bytecode (#35)
sampsyo/cs6120: content/blog/2019-09-25-bril2jb.md

+++
title = "bril2jb: A tool that translates bril to Java Bytecode"
extra.author = "Siqiu Yao"
extra.bio = """

>   [Siqiu Yao](http://www.cs.cornell.edu/~yaosiqiu/) is a 3rd year Ph.D. student interested in security about systems and programming languages.~~~~~~.

Reviewer: What’s with the tildes?

"""
+++

### Goal

> The goal of this project is to provide a tool to translate bril code to Java Bytecode.

Reviewer: ```suggestion
The goal of this project is to provide a tool to translate Bril code to Java bytecode.
```

Reviewer: Please capitalize Bril throughout your post.

### Design
#### Before the actual translation process

> At first glance, translating bril to [Java Bytecode](https://en.wikipedia.org/wiki/Java_bytecode) seems a 

Reviewer: Bytecode is lowercase, as that Wikipedia link suggests.

pretty straightforward job, since bril (currently) only contains simple instructions and supports no function invocation.

> But it turns out that making something directly runnable by *Java Virtual Machine*(JVM, the run-time environment of Java Bytecode)

Reviewer: Missing space.
```suggestion
But it turns out that making something directly runnable by *Java Virtual Machine* (JVM, the run-time environment of Java Bytecode)
```

is not trivial, because we need to construct our program as a Java *class*, 
which is the only code format JVM accepts.

> Therefore we need to create a wrapper class(say `Wrapper.class`) like the following :

Reviewer: ```suggestion
Therefore we need to create a wrapper class (say `Wrapper.class`) like the following :
```

Please be careful about spaces and parentheses throughout.

Reviewer: Why say Wrapper.class and not just Wrapper, which is the name of the class?

```JAVA
public class Wrapper {
  public Wrapper();
  public static void main(java.lang.String[]);

> }

Reviewer: Perhaps this is obvious, but this Java code does not parse (because the method bodies are missing).

```

> And the `main` function would contain the instructions in the bril code.

Reviewer: It’s a good point that you need to create a class, but I think this idea would be easier to understand if we knew how you were creating stuff like classes. So maybe this should come after you introduce ASM.

#### The translation process
I chose [ASM](https://asm.ow2.io), a Java Bytecode manipulation and analysis framework to help with the translation.

> The translation process is pretty regular. One interesting point is that Java Bytecode is strictly-typed, 

Reviewer: Java has both a stack and local variables. Which did you use for Bril variables?

I think it would help a lot if you gave an example—an input Bril program and the output JVM instructions.

Author: Do you mean the "operand stack" in stack frames or the overall stack that contains stack frames? I use local variables, which are contained in a stack frame.
I will explain a little bit on this in the new draft

Reviewer: Yep, the operand stack.

while bril is naturally dynamic and 
there might be no explicit declaration of a variable before it is used(in the order of instruction list),

> and some instructions are polymorphic such as `print`.

Reviewer: I disagree that Bril is “naturally dynamic.” There are no variable declarations, it’s true, but it’s statically and strictly typed.

This would perhaps be a good place to discuss your design decision (if you made one) about how to deal with programs that use different types for the same variable name. I know this came up on Slack, but the world doesn’t know that. :)

Reviewer: You say “some instructions” are polymorphic, but is anything other that “print” actually polymorphic? Please be precise about this.

Therefore, we preprocess the instructions first to gather all type information and labels.

> This step also infers types of variables, so there is actually no need to indicate types in bril anymore.

Reviewer: I’m not quite sure what you mean by this. What do you mean by saying you “inferred” the types? Did you just look for the assignments to the variables and use their types as written in the instructions that assign to them? (That’s not really inference; it’s just writing down the types.) Or did you need something more global?

Please expand this discussion with more detail.

  

### Implementation

> This tool is implemented in Java, and therefore maintained in a standalone repo. It translates bril code(in `json` format) to 

Reviewer: That’s not necessary! The main Bril repository already contains multiple languages, and it will contain more in the future.

Author: I'll merge it into Bril!

> This tool is implemented in Java, and therefore maintained in a standalone repo. It translates bril code(in `json` format) to 

Reviewer: ```suggestion
This tool is implemented in Java, and therefore maintained in a standalone repo. It translates bril code (in JSON format) to 
```

JSON is an acronym, not code.

> a same-name Java class file, which can run on JVM. 

Reviewer: ```suggestion
a same-name Java class file, which can run on the JVM. 
```

> For the usage guidance of our tool, please read `README.md` in [our repo](https://github.com/Neroysq/bril2jb).

Reviewer: To make your blog post self contained, I think it would be cool to include an interaction example right here instead of just delegating that to your README.

We assume that the input program is valid and has no faulty behaviors 

> because we think error-handling is not the topic of this project.   

Reviewer: What do you mean by “faulty behavior”? And what do you *think* might happen if your tool gets erroneous input? Will it crash, emit code that will fail the JVM verifier, or emit code that behaves arbitrarily at run time? Some examples could help here.

There are some tricky details:

1. Since `int` in bril is 64-bit, all `int` variables are `long` variables in JVM; 
there is no boolean type in JVM, so all `bool` variables are `int`(32-bit integer) type in JVM.

2. ASM handles the max local variable size,

>  but the index of local variable needs to be manually maintained(`long` type takes 2 index units while `int` takes 1). 

Reviewer: Please explain what the “max variable size” is for readers who are not familiar with JVM details.

>  but the index of local variable needs to be manually maintained(`long` type takes 2 index units while `int` takes 1). 

Reviewer: ```suggestion
 but the indices of local variables need to be manually maintained(`long` type takes 2 index units while `int` takes 1). 
```

 
3. The hardest part of this project, in my opinion, is translating `print`. 
The `print` instruction is translated as `java.lang.System.println`, 
and all arguments are converted to `String` then concatenated. 

> So before invoking the print function, we need to invoke the built-in concatenation function (`java.lang.invoke.StringConcatFactory.makeConcatWithConstants​`) first,

Reviewer: Comma splice. (Start a new sentence here.)

this process involves dynamic method invocation in Java, 
which requires a static Bootstrap Method to create it dynamically and 
therefore supports dynamic descriptors and arguments formatter 

> ([further reading](https://www.guardsquare.com/en/blog/string-concatenation-java-9-untangling-invokedynamic)).

Reviewer: Please include a little more detail. The phrases “bootstrap method” and “dynamic descriptors” and “arguments formatter” will be unfamiliar to your audience.

We need to generate a descriptor and formatter for each usage of `print`.

### Evaluation

> We assume all programs are valid bril programs.

Reviewer: Again, what exactly does this mean?

We tested our tool using manually-written test cases(including the ones in bril repo) to ensure the correctness.

> All of our tests pass and randomly chosen output Java Bytecode to look reasonable.

Reviewer: What does it mean for a test to “pass”? Did you manually look at the output, or do you have some way to check the results automatically?

Also, how many tests total did you run?

### Conclusion
In conclusion, we successfully built a translator from bril to Java bytecode. 
I look forward to further maintaining this tool to support potential new features 

> such as function call and memory allocation.

Reviewer: ```suggestion
such as function calls and memory allocation.
```
